### PR TITLE
ci: add workflow for checkrefs.py and typos

### LINKED ISSUE #105: Workflow for checking the mod
### Description
- [checkrefs.py](https://github.com/0ad/0ad/blob/master/source/tools/entity/checkrefs.py) is a good tool for testing the following:
  - check maps for missing actor and templates
  - verify actors files have no missing textures
  - check xml files match their (.rng) grammar file
- it would be beneficial if any kind of change to this mod would trigger the `checkrefs.py` script
  - best via a GitHub action workflow

#### Manual setup
- Currently, the user must verify the mod by first checking out 0 A.D. in the current release version.

```sh
# A26 current release tag
git restore --source A26 -SW < Path to 0 A.D. folder >
```

- navigate to the checkrefs script folder

```sh
cd < Path to 0 A.D. folder >/source/tools/entity
```

- run the script with the path to the `community-maps-2` mod and with the `-tax` option, see [readme.md](https://github.com/0ad/0ad/blob/master/source/tools/entity/readme.md) for explaination

```sh
python3 checkrefs.py -m < Path to community-maps-2 > -tax
INFO - Checking /Users/paria/Developer/community-maps-2's integrity.
# INFO - The following mods will be loaded: /Users/paria/Developer/community-maps-2|public|mod.
# INFO - Loading maps XML...
# INFO - Loading maps PMP...
# INFO - Loading entities...
# INFO - Loading actors...
# INFO - Loading variants...
# INFO - Loading art files...
# INFO - Loading materials...
# INFO - Loading particles...
# INFO - Loading sound groups...
# INFO - Loading audio files...
# INFO - Loading GUI XML...
# INFO - Loading GUI data...
# INFO - Loading civs...
# INFO - Loading random maps...
# INFO - Loading techs...
# INFO - Loading terrains...
# INFO - Loading auras...
# INFO - Loading tips...
# INFO - Looking for missing files...
# WARNING - Missing file 'art/actors/structures/hellenes/forge.xml' referenced by: /Users/paria/Developer/community-maps-2/maps/skirmishes/Lesvos Castle (2).xml
# WARNING - Missing file 'art/actors/structures/hellenes/sentry_tower.xml' referenced by: /Users/paria/Developer/community-maps-2/maps/skirmishes/Lesvos Castle (2).xml
# INFO - Validating actors...
...
```

---

#### Possible additional steps for Github actions
- check for typos, see #84

### PR DESCRIPTION
**Pull request by LangLangBart**
Fix #105

### Description
- 1st job
  - checkout 0ad repo at the latest release tag
  - run the checkrefs python script with the community-maps-2 mod
  - check for warnings/errors
- 2nd job
  - check for typos
  - the exceptions are specified in `.github/typos.toml`

---

PS: A normal run takes ~3-5min, because of the size of 0 A.D.

### CODE REVIEW

**Comment by LangLangBart** [comment]
Fixed the whitespace in `carpathian.js` while checking for typos.

**Comment by LangLangBart** [comment]
- The 1st job fails currently because not all issues are fixed.
- The lines with `- uses: hmarr/debug-action@v2`  is redundant, but won't hurt.

**Comment by LangLangBart** [comment]
Just tested the [GITHUB_STEP_SUMMARY](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary) feature,
 - the benefit with `>> GITHUB_STEP_SUMMARY` you can list all the issues and errors in one report (markdown format),
 -  with `echo "::warning:: {msg}"` one is limited to 10 warning annotations and 10 error annotations per step.

### Image with GITHUB_STEP_SUMMARY 
Screenshot 2023-03-01 at 14 05 19



### Image with Annotations 
Screenshot 2023-03-01 at 14 07 16


---
## Conclusion
I like the `annotations` better than  the `GITHUB_STEP_SUMMARY` you have to invest time to make it prettier and additionally that bloats the code unnecessarily. I will remove the TODO note from the `.yml` file.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+# Run the checkrefs.py script and check for typos.
+on:
+  push:
+    branches: [master]
```
Comment (andy5995) [inline .github/workflows/ci.yml:5]: Path filters needed here so the workflow will only run when relevant files are changed

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Get latest release tag of 0 A.D.
+        id: tag_name
+        run: |
+          echo "release_tag=$(curl -s https://api.github.com/repos/0ad/0ad/releases/latest | jq -r '.tag_name')" >> $GITHUB_OUTPUT
+      - name: Check out 0 A.D. (${{ steps.tag_name.outputs.release_tag }})
+        uses: actions/checkout@v3
+        with:
+          repository: 0ad/0ad
```
Comment (andy5995) [inline .github/workflows/ci.yml:21]: Instead, let's download the 0ad tar.gz (https://github.com/0ad/0ad/tags) and implement [caching](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#managing-caches)

There's a small example at https://github.com/curl/curl/pull/10393/files that fits our situation perfectly.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        run: |-
+          cd "$GITHUB_WORKSPACE/0ad_directory/source/tools/entity" || exit 1
+          array=()
+          while IFS='' read -r line; do array+=("$line"); done < <(python checkrefs.py -m "$GITHUB_WORKSPACE/community_maps_2_directory" -tax 2>&1)
+          for i in "${array[@]}"; do
+            [[ $i =~ ^WARNING ]] && echo "::warning:: $i"
+            [[ $i =~ ^ERROR ]] && echo "::error:: $i"
+          done
+          [[ "${array[*]}" =~ (WARNING|ERROR) ]] && exit 1 || exit 0
+  typos:
```
Comment (andy5995) [inline .github/workflows/ci.yml:43]: Let's not include this. At this time,  I don't think this action is very necessary for this type of repository.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          path: community_maps_2_directory
+      - uses: actions/setup-python@v4
+        with:
+          python-version: 3.x
+      - run: pip install lxml
+      - name: Run Python script
+        run: |-
+          cd "$GITHUB_WORKSPACE/0ad_directory/source/tools/entity" || exit 1
+          array=()
+          while IFS='' read -r line; do array+=("$line"); done < <(python checkrefs.py -m "$GITHUB_WORKSPACE/community_maps_2_directory" -tax 2>&1)
```
Comment (andy5995) [inline .github/workflows/ci.yml:37]: I like your solution to getting the required info. It's too bad it's such a lot of clutter to add to a workflow, but the only alternative I can think of: what do you think about making a patch file for the checkrefs.py script? The patch would be added to this repo, then applied after the 0ad source and data is retrieved...

And actually, the caching (mentioned above) could happen after the patch is applied.

**Review by andy5995 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (LangLangBart) [reply .github/workflows/ci.yml:21]: With caching, I got very similar speeds of around 3~5mins.
SCR-20230302-rbpd

The speed of getting the cache was almost the same as a new checkout:
SCR-20230302-rcos

```yml
    steps:
      - name: Get latest release tag of 0 A.D.
        id: tag_name
        run: |
          echo "release_tag=$(curl -s https://api.github.com/repos/0ad/0ad/releases/latest | jq -r '.tag_name')" >> $GITHUB_OUTPUT
      - name: Check if there is a cached version of 0 A.D. (${{ steps.tag_name.outputs.release_tag}})
        id: cache-0ad_directory
        uses: actions/cache@v3
        with:
          path: 0ad_directory
          key: ${{ runner.os }}-0ad_directory-${{ steps.tag_name.outputs.release_tag}}
      - name: Check out 0 A.D. if there is no cached version
        if: steps.cache-0ad_directory.outputs.cache-hit != 'true'
        uses: actions/checkout@v3
        with:
          repository: 0ad/0ad
          path: 0ad_directory
          ref: ${{ steps.tag_name.outputs.release_tag }}
          clean: true
```

Try to beat 3-5mins!

File: .github/workflows/ci.yml (hunk 4 above)
Comment (LangLangBart) [reply .github/workflows/ci.yml:37]: 😆

**Comment by LangLangBart** [comment]
You have got this.

**Pull request closed without merging**